File: iosApp/iosApp/Components/SignalSheet.swift
import SwiftUI

struct SignalPopup: View {
    let postId: String
    let typeKey: String
    var onSignalsUpdated: (Int) -> Void = { _ in }
    var onDismiss: () -> Void

    @Environment(\.appColors) private var colors
    @EnvironmentObject var lang: LanguageManager
    @State private var signalKeys: [SignalKey] = []
    @State private var signalGroups: [SignalGroup] = []
    @State private var postSignals: PostSignals?
    @State private var selectedKeys: Set<String> = []
    @State private var isLoading = true
    @State private var unsupportedType = false
    @State private var sheetY: CGFloat = UIScreen.main.bounds.height
    @State private var backdropOpacity: Double = 0
    @State private var dragOffset: CGFloat = 0

    private var hasChanges: Bool {
        selectedKeys != Set(postSignals?.mySignals ?? [])
    }

    var body: some View {
        ZStack(alignment: .bottom) {
            Color.black.opacity(backdropOpacity)
                .ignoresSafeArea()
                .onTapGesture { dismiss() }

            sheetContent
                .offset(y: sheetY + max(dragOffset, 0))
                .gesture(
                    DragGesture()
                        .onChanged { value in
                            dragOffset = value.translation.height
                        }
                        .onEnded { value in
                            if value.translation.height > 120 || value.velocity.height > 500 {
                                dismiss()
                            } else {
                                withAnimation(.interpolatingSpring(stiffness: 300, damping: 30)) {
                                    dragOffset = 0
                                }
                            }
                        }
                )
        }
        .ignoresSafeArea(edges: .bottom)
        .task {
            await loadData()
            // Data loaded — now animate in
            withAnimation(.interpolatingSpring(stiffness: 350, damping: 28)) {
                sheetY = 0
                backdropOpacity = 0.5
            }
        }
    }

    // Sheet is ALWAYS built — never conditionally removed/added
    private var sheetContent: some View {
        VStack(spacing: 0) {
            // Drag handle
            Capsule()
                .fill(colors.divider)
                .frame(width: 36, height: 5)
                .padding(.top, 10)
                .padding(.bottom, 16)

            // Header
            VStack(alignment: .leading, spacing: 8) {
                Text(lang.s("signal_title"))
                    .font(.system(size: 18, weight: .bold))
                    .foregroundColor(colors.textPrimary)
                Text(lang.s("signal_subtitle"))
                    .font(.system(size: 13))
                    .foregroundColor(colors.textSecondary)
                    .lineSpacing(2)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 20)
            .padding(.bottom, 20)

            if isLoading {
                ProgressView().tint(colors.accent)
                    .frame(height: 200)
                    .frame(maxWidth: .infinity)
            } else if unsupportedType {
                VStack(spacing: 12) {
                    Image(systemName: "exclamationmark.triangle")
                        .font(.system(size: 32))
                        .foregroundColor(colors.textSecondary.opacity(0.5))
                    Text(lang.s("signal_unsupported"))
                        .font(.system(size: 14))
                        .foregroundColor(colors.textSecondary)
                        .multilineTextAlignment(.center)
                }
                .frame(height: 160)
                .frame(maxWidth: .infinity)
                .padding(.horizontal, 20)
            } else {
                ScrollView(showsIndicators: false) {
                    VStack(spacing: 24) {
                        ForEach(signalGroups) { group in
                            SignalGroupView(
                                group: group,
                                selectedKeys: $selectedKeys,
                                signalCounts: postSignals?.signals ?? [:]
                            )
                        }
                    }
                    .padding(.horizontal, 20)
                    .padding(.bottom, 8)
                }
                .frame(maxHeight: UIScreen.main.bounds.height * 0.75)
                .fixedSize(horizontal: false, vertical: true)
            }

            Spacer().frame(height: 34)
        }
        .background(
            UnevenRoundedRectangle(topLeadingRadius: 24, topTrailingRadius: 24)
                .fill(colors.surface)
                .shadow(color: .black.opacity(0.3), radius: 20, y: -5)
        )
    }

    private func dismiss() {
        // Animate out
        withAnimation(.interpolatingSpring(stiffness: 400, damping: 30)) {
            sheetY = UIScreen.main.bounds.height
            backdropOpacity = 0
            dragOffset = 0
        }

        // Sync with API if there are changes
        if hasChanges {
            // Optimistic count: apply delta to current total
            let oldKeys = Set(postSignals?.mySignals ?? [])
            let added = selectedKeys.subtracting(oldKeys).count
            let removed = oldKeys.subtracting(selectedKeys).count
            let currentTotal = postSignals?.signals.values.reduce(0, +) ?? 0
            let previousTotal = currentTotal
            let optimisticTotal = max(currentTotal + added - removed, 0)
            onSignalsUpdated(optimisticTotal)

            let keys = Array(selectedKeys)
            let pid = postId
            let callback = onSignalsUpdated
            Task {
                do {
                    let response = try await ApiClient.shared.syncSignals(postId: pid, signalKeys: keys)
                    let serverTotal = response.signals.values.reduce(0, +)
                    await MainActor.run { callback(serverTotal) }
                } catch {
                    // Revert optimistic update and show error
                    await MainActor.run {
                        callback(previousTotal)
                        ToastManager.shared.show(LanguageManager.shared.s("toast_signal_error"))
                    }
                }
            }
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
            onDismiss()
        }
    }

    private func loadData() async {
        isLoading = true
        if let keys = await SignalKeysCache.shared.getKeys(for: typeKey) {
            signalKeys = keys
            signalGroups = SignalGroup.groupFromKeys(keys)
            postSignals = try? await ApiClient.shared.getPostSignals(postId: postId)
            if let existing = postSignals?.mySignals {
                selectedKeys = Set(existing)
            }
        } else {
            unsupportedType = true
        }
        isLoading = false
    }
}

// MARK: - Signal Group View

struct SignalGroupView: View {
    let group: SignalGroup
    @Binding var selectedKeys: Set<String>
    let signalCounts: [String: Int]

    @Environment(\.appColors) private var colors
    @EnvironmentObject var lang: LanguageManager

    private var localizedLabel: String {
        switch group.category {
        case "verification": return lang.s("signal_section_verification")
        case "reaction": return lang.s("signal_section_reaction")
        default: return group.label
        }
    }

    var body: some View {
        VStack(spacing: 14) {
            HStack(spacing: 8) {
                Rectangle()
                    .fill(colors.divider)
                    .frame(height: 0.5)
                Text(localizedLabel.uppercased())
                    .font(.system(size: 10, weight: .semibold))
                    .foregroundColor(colors.textSecondary)
                    .tracking(1)
                    .fixedSize()
                Rectangle()
                    .fill(colors.divider)
                    .frame(height: 0.5)
            }

            ForEach(group.pairs) { pair in
                HStack(spacing: 16) {
                    SignalItemView(
                        signal: pair.left,
                        isPositive: true,
                        isSelected: selectedKeys.contains(pair.left.key),
                        count: signalCounts[pair.left.key] ?? 0,
                        onTap: { toggleSignal(pair.left, opposite: pair.right) }
                    )

                    if pair.left.key != pair.right.key {
                        SignalItemView(
                            signal: pair.right,
                            isPositive: false,
                            isSelected: selectedKeys.contains(pair.right.key),
                            count: signalCounts[pair.right.key] ?? 0,
                            onTap: { toggleSignal(pair.right, opposite: pair.left) }
                        )
                    }
                }
            }
        }
    }

    private func toggleSignal(_ signal: SignalKey, opposite: SignalKey) {
        withAnimation(.easeInOut(duration: 0.2)) {
            if selectedKeys.contains(signal.key) {
                selectedKeys.remove(signal.key)
            } else {
                selectedKeys.insert(signal.key)
                if signal.key != opposite.key {
                    selectedKeys.remove(opposite.key)
                }
            }
        }
    }
}

// MARK: - Signal Item (circle style)

private let positiveColor = Color(red: 0.2, green: 0.7, blue: 0.5)
private let negativeColor = Color(red: 0.85, green: 0.35, blue: 0.35)

struct SignalItemView: View {
    let signal: SignalKey
    let isPositive: Bool
    let isSelected: Bool
    let count: Int
    let onTap: () -> Void

    @Environment(\.appColors) private var colors

    private var tintColor: Color {
        isPositive ? positiveColor : negativeColor
    }

    var body: some View {
        Button(action: onTap) {
            VStack(spacing: 6) {
                ZStack {
                    Circle()
                        .fill(isSelected ? tintColor.opacity(0.2) : tintColor.opacity(0.06))
                        .frame(width: 52, height: 52)

                    Circle()
                        .stroke(isSelected ? tintColor.opacity(0.5) : tintColor.opacity(0.12), lineWidth: 1.5)
                        .frame(width: 52, height: 52)

                    Text(emojiForSignal(signal.key))
                        .font(.system(size: 24))
                        .saturation(isSelected ? 1 : 0)
                        .scaleEffect(isSelected ? 1.1 : 1.0)

                    if count > 0 {
                        Text(formatCompactNumber(count))
                            .font(.system(size: 11, weight: .medium))
                            .tracking(0.3)
                            .foregroundColor(.white)
                            .padding(.horizontal, 7)
                            .padding(.vertical, 3)
                            .background(Capsule().fill(isSelected ? tintColor : colors.divider))
                            .offset(x: 22, y: -22)
                    }
                }

                Text(signal.label)
                    .font(.system(size: 10, weight: isSelected ? .semibold : .regular))
                    .foregroundColor(isSelected ? tintColor : colors.textSecondary)
                    .multilineTextAlignment(.center)
                    .lineLimit(2)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .frame(maxWidth: .infinity)
        }
        .buttonStyle(.plain)
    }
}

// MARK: - Emoji mapping

private func emojiForSignal(_ key: String) -> String {
    let map: [String: String] = [
        "saw_it": "👀",
        "i_confirm": "✅",
        "is_recurring": "🔁",
        "didnt_see": "🤷",
        "not_true": "🚫",
        "exaggerated": "😤",
        "solidarity": "🤝",
        "outrage": "😡",
        "fear": "😨",
        "relief": "😮‍💨",
        "gratitude": "🙏",
        "still_happening": "⏳",
        "resolved": "🔧",
        "me_too": "🙋",
        "urgent": "🔥",
        "helpful": "💡",
        "funny": "😂",
        "sad": "😢",
        "love": "❤️",
        "danger": "⚠️",
        "safe": "🛡️",
    ]
    return map[key] ?? "📌"
}

File: androidApp/src/main/kotlin/com/example/app/android/components/SignalSheet.kt
package com.example.app.android.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.app.android.R
import com.example.app.android.network.ApiClient
import com.example.app.android.network.SignalKeysCache
import com.example.app.android.network.models.PostSignals
import com.example.app.android.network.models.SignalGroup
import com.example.app.android.network.models.SignalKey
import com.example.app.android.network.models.SignalPair
import com.example.app.android.network.models.formatCompactNumber
import com.example.app.android.theme.AppTheme
import kotlinx.coroutines.launch

private val positiveColor = Color(0xFF33B380)
private val negativeColor = Color(0xFFD95959)

private val emojiMap = mapOf(
    "saw_it" to "\uD83D\uDC40",
    "i_confirm" to "\u2705",
    "is_recurring" to "\uD83D\uDD01",
    "didnt_see" to "\uD83E\uDD37",
    "not_true" to "\uD83D\uDEAB",
    "exaggerated" to "\uD83D\uDE24",
    "solidarity" to "\uD83E\uDD1D",
    "outrage" to "\uD83D\uDE21",
    "fear" to "\uD83D\uDE28",
    "relief" to "\uD83D\uDE2E\u200D\uD83D\uDCA8",
    "gratitude" to "\uD83D\uDE4F",
    "still_happening" to "\u231B",
    "resolved" to "\uD83D\uDD27",
    "me_too" to "\uD83D\uDE4B",
    "urgent" to "\uD83D\uDD25",
    "helpful" to "\uD83D\uDCA1",
    "funny" to "\uD83D\uDE02",
    "sad" to "\uD83D\uDE22",
    "love" to "\u2764\uFE0F",
    "danger" to "\u26A0\uFE0F",
    "safe" to "\uD83D\uDEE1\uFE0F"
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SignalSheet(
    postId: String,
    typeKey: String,
    onSignalsUpdated: (Int) -> Unit = {},
    onDismiss: () -> Unit
) {
    val colors = AppTheme.colors
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    var signalKeys by remember { mutableStateOf<List<SignalKey>?>(null) }
    var signalGroups by remember { mutableStateOf<List<SignalGroup>>(emptyList()) }
    var postSignals by remember { mutableStateOf<PostSignals?>(null) }
    var selectedKeys by remember { mutableStateOf(setOf<String>()) }
    var isLoading by remember { mutableStateOf(true) }
    var unsupportedType by remember { mutableStateOf(false) }

    // Remember initial state for change detection
    var initialKeys by remember { mutableStateOf(setOf<String>()) }

    LaunchedEffect(postId) {
        isLoading = true
        val keys = SignalKeysCache.getKeys(typeKey)
        if (keys != null) {
            signalKeys = keys
            signalGroups = SignalGroup.groupFromKeys(keys)
            ApiClient.getPostSignals(postId).onSuccess { ps ->
                postSignals = ps
                selectedKeys = ps.mySignals.toSet()
                initialKeys = ps.mySignals.toSet()
            }
        } else {
            unsupportedType = true
        }
        isLoading = false
    }

    fun syncAndDismiss() {
        val hasChanges = selectedKeys != initialKeys
        if (hasChanges) {
            // Optimistic count
            val oldKeys = initialKeys
            val added = selectedKeys.subtract(oldKeys).size
            val removed = oldKeys.subtract(selectedKeys).size
            val currentTotal = postSignals?.signals?.values?.sum() ?: 0
            val previousTotal = currentTotal
            val optimisticTotal = maxOf(currentTotal + added - removed, 0)
            onSignalsUpdated(optimisticTotal)

            val keys = selectedKeys.toList()
            scope.launch {
                ApiClient.syncSignals(postId, keys).fold(
                    onSuccess = { response ->
                        val serverTotal = response.signals.values.sum()
                        onSignalsUpdated(serverTotal)
                    },
                    onFailure = {
                        onSignalsUpdated(previousTotal)
                        ToastManager.show(context.getString(R.string.toast_signal_error))
                    }
                )
            }
        }
        onDismiss()
    }

    ModalBottomSheet(
        onDismissRequest = { syncAndDismiss() },
        sheetState = sheetState,
        containerColor = colors.surface,
        dragHandle = {
            Box(
                modifier = Modifier
                    .padding(top = 10.dp, bottom = 16.dp)
                    .size(width = 36.dp, height = 5.dp)
                    .clip(RoundedCornerShape(50))
                    .background(colors.divider)
            )
        }
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(bottom = 34.dp)
        ) {
            // Header
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 20.dp)
                    .padding(bottom = 20.dp)
            ) {
                Text(
                    text = context.getString(R.string.signal_title),
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    color = colors.textPrimary
                )
                Spacer(Modifier.height(8.dp))
                Text(
                    text = context.getString(R.string.signal_subtitle),
                    fontSize = 13.sp,
                    color = colors.textSecondary,
                    lineHeight = 18.sp
                )
            }

            when {
                isLoading -> {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(200.dp),
                        contentAlignment = Alignment.Center
                    ) {
                        CircularProgressIndicator(color = colors.accent, strokeWidth = 2.dp, modifier = Modifier.size(24.dp))
                    }
                }
                unsupportedType -> {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(160.dp)
                            .padding(horizontal = 20.dp),
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.Center
                    ) {
                        Icon(
                            Icons.Default.Warning,
                            contentDescription = null,
                            tint = colors.textSecondary.copy(alpha = 0.5f),
                            modifier = Modifier.size(32.dp)
                        )
                        Spacer(Modifier.height(12.dp))
                        Text(
                            text = context.getString(R.string.signal_unsupported),
                            fontSize = 14.sp,
                            color = colors.textSecondary,
                            textAlign = TextAlign.Center
                        )
                    }
                }
                else -> {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .verticalScroll(rememberScrollState())
                            .padding(horizontal = 20.dp)
                            .padding(bottom = 8.dp),
                        verticalArrangement = Arrangement.spacedBy(24.dp)
                    ) {
                        signalGroups.forEach { group ->
                            SignalGroupSection(
                                group = group,
                                selectedKeys = selectedKeys,
                                signalCounts = postSignals?.signals ?: emptyMap(),
                                context = context,
                                onToggle = { signal, opposite ->
                                    selectedKeys = if (signal.key in selectedKeys) {
                                        selectedKeys - signal.key
                                    } else {
                                        val newSet = selectedKeys + signal.key
                                        if (signal.key != opposite.key) newSet - opposite.key else newSet
                                    }
                                }
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun SignalGroupSection(
    group: SignalGroup,
    selectedKeys: Set<String>,
    signalCounts: Map<String, Int>,
    context: android.content.Context,
    onToggle: (SignalKey, SignalKey) -> Unit
) {
    val colors = AppTheme.colors
    val localizedLabel = when (group.category) {
        "verification" -> context.getString(R.string.signal_section_verification)
        "reaction" -> context.getString(R.string.signal_section_reaction)
        else -> group.label
    }

    Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
        // Category header with lines
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Box(Modifier.weight(1f).height(0.5.dp).background(colors.divider))
            Text(
                text = localizedLabel.uppercase(),
                fontSize = 10.sp,
                fontWeight = FontWeight.SemiBold,
                color = colors.textSecondary,
                letterSpacing = 1.sp,
                modifier = Modifier.padding(horizontal = 8.dp)
            )
            Box(Modifier.weight(1f).height(0.5.dp).background(colors.divider))
        }

        // Signal pairs
        group.pairs.forEach { pair ->
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(16.dp)
            ) {
                Box(Modifier.weight(1f)) {
                    SignalItem(
                        signal = pair.left,
                        isPositive = true,
                        isSelected = pair.left.key in selectedKeys,
                        count = signalCounts[pair.left.key] ?: 0,
                        onClick = { onToggle(pair.left, pair.right) }
                    )
                }
                if (pair.left.key != pair.right.key) {
                    Box(Modifier.weight(1f)) {
                        SignalItem(
                            signal = pair.right,
                            isPositive = false,
                            isSelected = pair.right.key in selectedKeys,
                            count = signalCounts[pair.right.key] ?: 0,
                            onClick = { onToggle(pair.right, pair.left) }
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun SignalItem(
    signal: SignalKey,
    isPositive: Boolean,
    isSelected: Boolean,
    count: Int,
    onClick: () -> Unit
) {
    val colors = AppTheme.colors
    val tintColor = if (isPositive) positiveColor else negativeColor
    val scale by animateFloatAsState(
        targetValue = if (isSelected) 1.1f else 1.0f,
        animationSpec = spring(dampingRatio = 0.6f),
        label = "scale"
    )

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(
                interactionSource = remember { MutableInteractionSource() },
                indication = null,
                onClick = onClick
            ),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(contentAlignment = Alignment.Center) {
            // Background circle
            Box(
                modifier = Modifier
                    .size(52.dp)
                    .clip(CircleShape)
                    .background(if (isSelected) tintColor.copy(alpha = 0.25f) else colors.divider.copy(alpha = 0.3f))
            )

            // Emoji
            Text(
                text = emojiMap[signal.key] ?: "\uD83D\uDCCC",
                fontSize = 24.sp,
                modifier = Modifier.scale(scale)
            )

            // Count badge
            if (count > 0) {
                Box(
                    modifier = Modifier
                        .align(Alignment.TopEnd)
                        .offset(x = 4.dp, y = (-4).dp)
                        .clip(RoundedCornerShape(50))
                        .background(if (isSelected) tintColor else colors.divider)
                        .padding(horizontal = 7.dp, vertical = 3.dp),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = formatCompactNumber(count),
                        fontSize = 11.sp,
                        fontWeight = FontWeight.Medium,
                        color = Color.White,
                        letterSpacing = 0.3.sp
                    )
                }
            }
        }

        Spacer(Modifier.height(6.dp))

        Text(
            text = signal.label,
            fontSize = 10.sp,
            fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
            color = if (isSelected) tintColor else colors.textSecondary,
            textAlign = TextAlign.Center,
            maxLines = 2,
            overflow = TextOverflow.Ellipsis
        )
    }
}
